search/search_repository.go

```diff
@@ -0,0 +1,361 @@
+
+	db := r.db.Model(&models.WorkItem{}).Where(sqlSearchQuery, sqlSearchQueryParameter)
+	orgDB := db
+	if start != nil {
+		if *start < 0 {
+			return nil, 0, BadParameterError{"start", *start}
+		}
+		db = db.Offset(*start)
+	}
+	if limit != nil {
+		if *limit <= 0 {
+			return nil, 0, BadParameterError{"limit", *limit}
+		}
+		db = db.Limit(*limit)
+	}
+	db = db.Select("count(*) over () as cnt2 , *")
+
+	rows, err := db.Rows()
+	if err != nil {
+		return nil, 0, err
+	}
+	defer rows.Close()
+
+	result := []models.WorkItem{}
+	value := models.WorkItem{}
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, 0, InternalError{simpleError{err.Error()}}
+	}
+
+	// need to set up a result for Scan() in order to extract total count.
+	var count uint64
+	var ignore interface{}
+	columnValues := make([]interface{}, len(columns))
+
+	for index := range columnValues {
+		columnValues[index] = &ignore
+	}
+	columnValues[0] = &count
+	first := true
+
+	for rows.Next() {
+		db.ScanRows(rows, &value)
+		if first {
+			first = false
+			if err = rows.Scan(columnValues...); err != nil {
+				return nil, 0, InternalError{simpleError{err.Error()}}
+			}
+		}
+		result = append(result, value)
+
+	}
+	if first {
+		// means 0 rows were returned from the first query (maybe becaus of offset outside of total count),
+		// need to do a count(*) to find out total
+		orgDB := orgDB.Select("count(*)")
+		rows2, err := orgDB.Rows()
+		defer rows2.Close()
+		if err != nil {
+			return nil, 0, err
+		}
+		rows2.Next() // count(*) will always return a row
+		rows2.Scan(&count)
+	}
+	return result, count, nil
+	//*/
+}
+
+// SearchFullText Search returns work items for the given query
+func (r *GormSearchRepository) SearchFullText(ctx context.Context, rawSearchString string, start *int, limit *int) ([]*app.WorkItem, uint64, error) {
+	// parse
+	// generateSearchQuery
+	// ....
+	parsedSearchDict := parseSearchString(rawSearchString)
+
+	sqlSearchQuery, sqlSearchQueryParameter := generateSQLSearchInfo(parsedSearchDict)
+	var rows []models.WorkItem
+	rows, count, err := r.search(ctx, sqlSearchQuery, sqlSearchQueryParameter, start, limit)
+	if err != nil {
+		return nil, 0, err
+	}
+	result := make([]*app.WorkItem, len(rows))
+
+	for index, value := range rows {
+		var err error
+		// FIXME: Against best practice http://go-database-sql.org/retrieving.html
+		wiType, err := r.loadTypeFromDB(ctx, value.Type)
+		if err != nil {
+			return nil, 0, InternalError{simpleError{err.Error()}}
+		}
+		result[index], err = convertFromModel(*wiType, value)
+		if err != nil {
+			return nil, 0, ConversionError{simpleError{err.Error()}}
+		}
+	}
+
+	return result, count, nil
+}
+
+// Validate ensures that the search string is valid and also ensures its not an injection attack.
+func (r *GormSearchRepository) Validate(ctx context.Context, rawSearchString string) error {
```

Comment:
Thanks @tsmaeder ,
Your argument sounds valid to me. I'll remove the Validate(..).
`	db := r.db.Model(&models.WorkItem{}).Where(sqlSearchQuery, sqlSearchQueryParameter)` 

Should be taking care of query injection issues.